gnovm/stdlibs/std/coins.gno

```diff
@@ -13,28 +14,147 @@ type Coin struct {
+		Amount: amount,
+	}
+}
+
+// String provides a human-readable representation of a coin
 func (c Coin) String() string {
+	if c.IsZero() {
+		return "0" + c.Denom
+	}
+
 	return strconv.Itoa(int(c.Amount)) + c.Denom
 }
 
+// IsGTE returns true if they are the same type and the receiver is
+// an equal or greater value
 func (c Coin) IsGTE(other Coin) bool {
 	if c.Denom != other.Denom {
-		panic("invalid coin denominations: " + c.Denom)
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
 	}
+
 	return c.Amount >= other.Amount
 }
 
+// IsLT returns true if they are the same type and the receiver is
+// a smaller value
+func (c Coin) IsLT(other Coin) bool {
+	if c.Denom != other.Denom {
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
+	}
+
+	return c.Amount < other.Amount
+}
+
+// IsEqual returns true if the two sets of Coins have the same value
+func (c Coin) IsEqual(other Coin) bool {
+	if c.Denom != other.Denom {
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
+	}
+
+	return c.Amount == other.Amount
+}
+
+// Adds amounts of two coins with same denom.
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Add(coinB Coin) Coin {
+	res := c.AddUnsafe(coinB)
+	return res
+}
+
+func (c Coin) AddUnsafe(coinB Coin) Coin {
+	if c.Denom != coinB.Denom {
+		panic("incompatible coin denominations: " + c.Denom + ", " + coinB.Denom)
+	}
+
+	sum, ok := overflow.Add64(c.Amount, coinB.Amount)
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+
+	return Coin{c.Denom, sum}
+}
+
+// Subtracts amounts of two coins with same denom.
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Sub(coinB Coin) Coin {
+	res := c.SubUnsafe(coinB)
+	return res
+}
+
+func (c Coin) SubUnsafe(coinB Coin) Coin {
+	if c.Denom != coinB.Denom {
+		panic("incompatible coin denominations: " + c.Denom + ", " + coinB.Denom)
+	}
+	dff, ok := overflow.Sub64(c.Amount, coinB.Amount)
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+	return Coin{c.Denom, dff}
+}
+
+// IsPositive returns true if coin amount is positive.
+func (c Coin) IsPositive() bool {
+	return c.Amount > 0
+}
+
+// IsNegative returns true if the coin amount is negative and false otherwise.
+func (c Coin) IsNegative() bool {
+	return c.Amount < 0
+}
+
+// IsZero returns if this represents no money
+func (c Coin) IsZero() bool {
+	return c.Amount == 0
+}
+
 // Coins is a set of Coin, one per currency
 type Coins []Coin
 
+func NewCoins(coins ...Coin) Coins {
+	return coins
+}
+
+// NewCoinsFromVal is a type-unsafe way to make a Coins struct
+// Usage is not recommended
+func NewCoinsFromVal(val ...interface{}) Coins {
```

Comment:
I think we don't need this constructor, it's bloat

Reply:
[22a171e](https://github.com/gnolang/gno/pull/2104/commits/22a171e5bbc244ee6c2d9b604960cc62862a9601)